### ISSUE TITLE
Adding constexpr to static const functions

### ISSUE THREAD

**Issue by ponast** [commschamp/commsdsl#16]
The name() function seems to always return a pointer to a compile time defined string. It has the following signature:
`static const char* name();`

I see no reason it should not be possible to add constexpr to the name functions other than backwards compatibility issues. Adding constexpr would enable the use of the function in compile time initialization of parameters in user code. Now it is not possible to do so which means duplication of name definitions. The same name thus must be defined in the schema file and in the c++ sources.

My intended use is as follows:
```
struct Xml_tags
{
    using Max_qos = mqtt5::field::SubscribeOptionsMembersCommon::MaxQosCommon;
    constexpr char static * max_qos = Max_qos::name();
    // More tags follows
};
```

**Comment by arobenko** [comment]
Hi @ponast,
I try not to abuse usage of `constexpr` in my work in general, and it CommsChampion Ecosystem in particular. Although C++ is quite flexible in allowing indiscriminate usage of this keyword and spits out errors only when the result of `constexpr` function is attempted to be used for compile time code generation conditions (such as using classes from `type_traits`), but evaluation of the return value at compile time cannot be calculated. 

My convention is to add `constexpr` **only** for functions that are **expected** to be used for compile time evaluation, mostly for boolean expressions of `std::conditional<...>` or similar. It becomes a "contract" for the function that it's usage of the compile time evaluation will never fail. Once `constexpr` - always `constexpr`. Anything else I leave for the regular compiler optimizations.

As for the suggested change, the `constexpr` on the return type of `const char*` doesn't make much sense, it doesn't preserve the size of the string, i.e. it's not `const char[N]`. To allow the compile time strings the name should be a static variable, rather than a function
```
static constexpr const char NameStr[] = "some_name";
```
However, my experiments show that some compilers (clang for example) don't handle this case properly. The link stage reports error of inability to find the string symbol:
```
/usr/bin/ld: /usr/bin/ld: DWARF error: invalid or unhandled FORM value: 0x23
...:(.text+0x14e2e): undefined reference to `SomeField::NameStr'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
However if the constexpr string is put in the anonymous namespace in the cpp file instead of the static member of the class, everything compiles normally.

Also worth mentioning that when developing initial versions of the code generator I intended to make sure that the strings are not duplicated, i.e. having single class accumulating all the strings to avoid repetition. However, when analyzing the binary code produced by the compiler I noticed that it cleans up all the duplicates by itself during its optimization stage, i.e. the following classes will reference the same and the only "SomeName" string in the read only section of the binary.
```
struct SomeField
{
    static const char* name()
    {
        return "SomeName";
    }
};

struct SomeOtherField
{
    static const char* name()
    {
        return "SomeName";
    }
};
```
As the result complicating the code for strings storage wasn't worth the effort.

Another thing: I personally use CommsChampion Ecosystem at work for the old product that is compiled using quite old gcc-4.8. So I put a significant effort not to introduce code supported by C++14/17/20, but not C++11 or having difficulties being compiled with gcc-4.8. At least not at this stage.

----

The bottom line, I don't think your suggestion of adding `constexpr` to the signature of `name()` member function has any (significant) added value.

**Comment by ponast** [comment]
Thanks for the thorough elaboration. You definitely have considered the topic in more depth than I have. I will take a look at my own use of constexpr which until now has been indiscriminate.
